src/custom/Doppelganger.sol

```diff
@@ -0,0 +1,173 @@
+        string newUri,
+        uint256 index
+    );
+
+    /// @notice Event emitted when a uri is cloned
+    event Cloned(
+        address indexed sender,
+        uint256 tokenId,
+        string newUri
+    );
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                    Error
+    //////////////////////////////////////////////////////////////////////////*/
+
+    error Unauthorized();
+
+    error MetadataSelectionDoesNotExist(uint256 selection);
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                    Structs
+    //////////////////////////////////////////////////////////////////////////*/
+
+    struct DoppelgangerStorage {
+        mapping(uint256 => uint256) dopplegangTokens;
+        string[] uris;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                    Constructor
+    //////////////////////////////////////////////////////////////////////////*/
+
+    /// @param name: the name of the contract
+    /// @param symbol: the symbol of the contract
+    /// @param defaultRoyaltyRecipient: the default address for royalty payments
+    /// @param defaultRoyaltyPercentage: the default royalty percentage of basis points (out of 10,000)
+    /// @param initOwner: initial owner of the contract
+    /// @param admins: array of admin addresses to add to the contract
+    /// @param enableStory: a bool deciding whether to add story fuctionality or not
+    /// @param blockListRegistry: address of the blocklist registry to use
+    constructor(
+        address implementation,
+        string memory name,
+        string memory symbol,
+        address defaultRoyaltyRecipient,
+        uint256 defaultRoyaltyPercentage,
+        address initOwner,
+        address[] memory admins,
+        string memory defaultUri,
+        bool enableStory,
+        address blockListRegistry
+    )
+        ERC1967Proxy(
+            implementation,
+            abi.encodeWithSelector(
+                0x1fbd2402, // selector for "initialize(string,string,address,uint256,address,address[],bool,address)"
+                name,
+                symbol,
+                defaultRoyaltyRecipient,
+                defaultRoyaltyPercentage,
+                initOwner,
+                admins,
+                enableStory,
+                blockListRegistry
+            )
+        )
+    {
+        DoppelgangerStorage storage store;
+
+        assembly {
+            store.slot := METADATA_STORAGE_SLOT
+        }
+
+        store.uris.push(defaultUri);
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+                                Admin Write Functions
+    //////////////////////////////////////////////////////////////////////////*/
+
+    function addDoppelganger(string calldata _newDoppelganger) external {
```

Comment:
depending on how many doppelgangers he has to start, we may want to think about making this a looped function @koloz193